### PR TITLE
Support hhmm timezone format

### LINKED ISSUE #43915: Support +HHMM timezone format in CSV upload
### Description

A common date format option `%z` prints the offset as 4 straight digits:

```python
datetime.now().astimezone().strftime('%Y-%m-%dT%H:%M:%S%z')

'2024-06-10T07:55:59+0000'
``` 

Unfortunately Metabase does not currently support this format, and will parse it as string instead of a timestamp.

It does however handle the very similar and slightly more common ISO 8601 format: `2024-06-10T07:55:59+00:00`

It would be great if Metabase could simply handle the original timestamp, to save customers needing to do further pre-processing.

Unfortunately, as Metabase defers the parsing to the Java standard library, it's not trivial to extend the supported pattern

### PR DESCRIPTION
**Pull request by duncanmak**
Closes https://github.com/metabase/metabase/issues/43915

### Description

Try a custom formatter if the default formatter for OffsetDateTime doesn't work

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by crisptrutski** [comment]
This looks reasonable to me. 

It's unfortunately that we need more parsing phases and exception handling, at some point it would be nice to collapse this all down to a single pass with a unified grammar.

Can you please just update [this test](https://github.com/metabase/metabase/blob/bab905c069c2215c04dda8f0df2a97d00a878c1b/test/metabase/upload_test.clj#L604) as well, so we have more end-to-end coverage?

**Comment by crisptrutski** [comment]
Thanks for the contribution!

**Review by crisptrutski — APPROVED** [review]

**Comment by duncanmak** [comment]
@crisptrutski is there anything else I need to do to get this PR merged?

I'm new to the repo and I see that some of the checks listed above are failing - is it because I forgot to update something, or are some of the checks known to fail?

Thanks for the quick review!

**Comment by crisptrutski** [comment]
I'm not sure why most of the tests failed, it looks like a CI problem.

I'm going to push a quick commit adding that private metadata and see if my elevated permissions help it build.

**Comment by crisptrutski** [comment]
OK so CI has run properly now 🙂 

Unfortunately there is a test that has been broken:

```
 FAIL in metabase.upload-test/detect-schema-offset-datetimes-test (upload_test.clj:237)

:h2
 Dates
expected: {:not_datetime :metabase.upload.types/varchar-255,
           :offset_datetime :metabase.upload.types/offset-datetime}
  actual: {:not_datetime :metabase.upload.types/offset-datetime,
           :offset_datetime :metabase.upload.types/offset-datetime}
```

From a quick look, it seems that leap days on non leap years are now being accepted by the new formatter:

```
               ["Offset Datetime,Not Datetime"
                "2022-01-01T00:00:00-01:00,2023-02-28T00:00:00-01:00"
                "2022-01-01T00:00:00-01:00,2023-02-29T00:00:00-01:00"
                "2022-01-01T00:00:00Z,2023-02-29T00:00:00-01:00"]))))))
```

**Comment by duncanmak** [comment]
@crisptrutski I'm still working on this, sorry, I've been busy with other things in the past few days. I'd like to get this PR merged by the end of this week.

**Comment by crisptrutski** [comment]
No problem - end of week will still be great!

**Comment by duncanmak** [comment]
@crisptrutski Sorry this took a while, I got it to pass all the tests locally, can you give this another look?

I'm not completely happy with how this worked out, but if it passes all the tests, maybe this PR is still good enough to go in?

I thought of a different approach: Use instaparse to extract the relevant values from the input string, and then use the OffsetDateTime/of constructor to create value. Doing so would avoid the multiple (somewhat specific) exception catch blocks, and the code will have a simpler control flow. 

I'm not sure such a small feature warrants writing a custom parser, but I do see that Instaparse if already in `deps.edn`.

----

I ran the following tests locally:

```
clojure -X:dev:test :only metabase.upload.types-test
clojure -X:dev:test :only metabase.upload-test
```

**Comment by crisptrutski** [comment]
Thanks for the perseverance here, but unfortunately this wasn't quite right yet. It turns out there was actually a bug in our tests, and when I fixed that this branch was still failing.

I did a bit of digging, and found that there's a "resolver" property on the formatter that can be configured to validate the actual date. For some reason the default implementation silently rounds down to the nearest real date, dubious 😄 

I also wasn't able to reproduce the same exception message that you were checking for. I think matching on just the type is fine, but I found myself getting  different exception to yours. Let's see what happens when CI runs, perhaps this is JDK version specific, but that seems unlikely.

Lastly, there was a bug in the exception handling - we would no longer be catching the other exception types around the initial parse attempt, so we would no longer be doing the expected substitution with our own custom exception.

Since the fixes to all of these were rather minor, I hope you don't mind that I've pushed the relevant tweaks. I'll still make sure that you get credit for this change in the final merge.

File: test/metabase/upload_test.clj [hunk 1]
```diff
@@ -236,14 +236,17 @@
...
 (deftest ^:parallel detect-schema-offset-datetimes-test
   (mt/test-drivers (mt/normal-drivers-with-feature :uploads)
-    (testing "Dates"
-      (is (=? {:offset_datetime offset-dt-type
-               :not_datetime   vchar-type}
-              (detect-schema-with-csv-rows
-               ["Offset Datetime,Not Datetime"
-                "2022-01-01T00:00:00-01:00,2023-02-28T00:00:00-01:00"
-                "2022-01-01T00:00:00-01:00,2023-02-29T00:00:00-01:00"
-                "2022-01-01T00:00:00Z,2023-02-29T00:00:00-01:00"]))))))
```
Comment (crisptrutski) [inline test/metabase/upload_test.clj:246]: The bug in the test was that either of these two rows was sufficient to reject the right hand column as not-a-date, leaving us free to infer the wrong type for one of them.

File: test/metabase/upload_test.clj [hunk 2]
```diff
@@ -236,14 +236,17 @@
...
+    (let [good-versus-bad-rows ["2022-01-01T00:00:00-01:00,2023-02-29T00:00:00-01:00"
+                                "2022-01-01T00:00:00-0100,2023-02-29T00:00:00-0100"
```
Comment (crisptrutski) [inline test/metabase/upload_test.clj:240]: I added this case to exercise the new format.

File: src/metabase/upload/parsing.clj [hunk 3]
```diff
@@ -105,6 +105,25 @@
...
+    (.appendPattern ".SSS")
+    (.optionalEnd)
+    (.optionalStart)
+    (.appendZoneOrOffsetId)
+    (.optionalEnd)
+    (.optionalStart)
+    (.appendOffset "+HHMM", "Z")
+    (.optionalEnd)
+    (.toFormatter)
+    (.withResolverStyle ResolverStyle/STRICT)))
```
Comment (crisptrutski) [inline src/metabase/upload/parsing.clj:125]: Here's the key bit

**Comment by crisptrutski** [comment]
@duncanmak The build is now passing. I've also rebased it on the latest master and added a final commit in your name so that you'll get the credit in the squashed commit. Let me know if you're happy for me to merge. Thanks again for this contribution!

**Comment by duncanmak** [comment]
Let's merge this in!

**Pull request merged**